storage/imxrt-flash/flashsrv.c

```diff
@@ -94,100 +345,405 @@ static void flashsrv_devCtl(msg_t *msg)
+			part->oid.port = flashsrv_common.flash_memories[part->fID].rawPort;
+
+			if ((res = create_dev(&part->oid, path)) < 0) {
+				LOG_ERROR("imxrt-flashsrv: create %s - err: %d", path, res);
+				return res;
+			}
+
+			if (!flashsrv_common.flash_memories[part->fID].rawActive) {
+				flashsrv_common.flash_memories[part->fID].rawActive = 1;
+				if ((mem = malloc(THREAD_STACKSZ)) == NULL) {
+					LOG_ERROR("imxrt-flashsrv: cannot alloc memory.");
+					return -ENOMEM;
+				}
+
+				beginthread(flashsrv_rawThread, THREAD_PRIORITY, mem, THREAD_STACKSZ, (void *)&flashsrv_common.flash_memories[part->fID]);
+			}
+			part->pStatus = flash_memory_active;
+			break;
+
+		/* Each meterfs partiton is handled by separete thread */
+		case ptable_meterfs:
+			if ((res = flashsrv_initMeterfs(part)) < 0)
+				return res;
+
+			portCreate(&part->oid.port);
+			portRegister(part->oid.port, path, NULL);
+			if ((res = create_dev(&part->oid, path)) < 0) {
+				LOG_ERROR("imxrt-flashsrv: create %s - err: %d", path, res);
+				return res;
+			}
+
+			if ((mem = malloc(THREAD_STACKSZ)) == NULL) {
+				LOG_ERROR("imxrt-flashsrv: cannot alloc memory.");
+				return -ENOMEM;
+			}
+
+			beginthread(flashsrv_meterfsThread, THREAD_PRIORITY, mem, THREAD_STACKSZ, (void *)part);
+			part->pStatus = flash_memory_active;
+			break;
+
+		default:
+			break;
+	}
+
+	return res;
+}
+
+
+static int flashsrv_partsInit(void)
+{
+	uint32_t ptabAddr;
+	uint32_t magicAddr;
+	int i, j, err = EOK;
+
+	flash_memory_t *memory;
+	ptable_header_t *tHeader;
+	ptable_partition_t *pHeader;
+
+	for (i = 0; i < FLASH_MEMORIES_NO; ++i) {
+		memory = flashsrv_common.flash_memories + i;
+		if (memory->fStatus != flash_memory_active)
+			continue;
+
+		ptabAddr = flashsrv_common.flexspi_adresses[i] + memory->ctx.properties.size - memory->ctx.properties.sector_size;
+		tHeader = (ptable_header_t *)(ptabAddr);
+		magicAddr = ptabAddr + sizeof(ptable_header_t) + tHeader->pSize * sizeof(ptable_partition_t);
+
+		/* Check values from partition table header */
+		if ((tHeader->pSize > (memory->ctx.properties.sector_size / (sizeof(ptable_partition_t) + sizeof(ptable_header_t) + sizeof(magicBytes)))) ||
+			(magicAddr > (flashsrv_common.flexspi_adresses[i] + memory->ctx.properties.size - sizeof(magicBytes))))
+			continue;
+
+		/* Verify magic bytes */
+		if ((memcmp((void *)(magicAddr), (void *)magicBytes, sizeof(magicBytes)) != 0)) {
+			LOG_ERROR("imxrt-flashsrv: partition table is not located on flash %d.", i);
+			continue;
+		}
+
+		if ((memory->parts = calloc(tHeader->pSize, sizeof(flashsrv_partition_t))) == NULL) {
+			LOG_ERROR("imxrt-flashsrv: cannot allocate memory.");
+			return -ENOMEM;
+		}
+
+		memory->pSz = tHeader->pSize;
+
+		/* Initialize partitions */
+		for (j = 0; j < tHeader->pSize; ++j) {
+			pHeader = (ptable_partition_t *)(ptabAddr + sizeof(ptable_header_t) + j * sizeof(ptable_partition_t));
+
+			if (pHeader->size < memory->ctx.properties.sector_size * MIN_PARTITIONS_SECTORS_NB) {
```

Comment:
Additional notes:
- we're not checking if partition lays on the flash itself (either `pHeader->offset` and `pHeader->offset + pHeader->size`). I don't know about meterfs, but for RAW partitions there are no additional checks during initialization (this should probably be added)
- we're not checking if the partitions overlap - I don't know if it's really needed to be checked on the device (would it introduce some hard-to-see errors?)